src/openvpn/mroute.c

```diff
@@ -223,30 +223,41 @@ mroute_extract_addr_ip(struct mroute_addr *src, struct mroute_addr *dest,
     return ret;
 }
 
+static void mroute_copy_ether_to_addr(struct mroute_addr *maddr,
+                                      const uint8_t *eth_addr,
+                                      uint16_t vid)
+{
+    maddr->type = MR_ADDR_ETHER;
+    maddr->netbits = 0;
+    memcpy (maddr->raw_addr, eth_addr, 6);
+#ifdef ENABLE_VLAN_TAGGING
+    maddr->len = 8;
+    memcpy (maddr->raw_addr + 6, &vid, 2);
```

Comment:
What happens if VLAN_TAGGING is compiled in (thus #ifdef ENABLE_VLAN_TAGGING is true), but it is not enabled at all in the config? do we still have the 802.1q header in the packet?

Reply:
as a side note...
why having those #ifdef at all ?
vlan does not require external libraries, neither it is too have for wrt devices.

less ifdefs --> less complexity

Reply:
@chipitsine have you measured the code footprint with and without VLAN_TAGGING? A few KB can make a big difference on LEDE devices.

This said, I agree that we should have throw ifdefs into the code. There is a much better and cleaner way to handle these cases (i.e. see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/batman-adv/distributed-arp-table.h?h=v4.14-rc1 with respect to the CONFIG_BATMAN_ADV_DAT ifdef) where every if-logic is contained in the header files.

However, I'd like to focus on the concept first and comment on the code later.

Reply:
I think this is just to extract the address information (and these patches co-opt it slightly to also return the vlan ID of the packet).  It appears to copy the vid each time regardless, which is then handed back out of the function.  Looking through the calls this primarily would seem to affect the `multi_bcast` method, but I think it could be resolved by adding an else clause to [line 285](https://github.com/OpenVPN/openvpn/pull/76/files/d9ee9f365ea49b9b843f06a36cf4e5249ca9798e#diff-cc63c55c1eee06c2b95fae210c506d4eR285) to set the vid back to 0 again if the packet is not of type 802.1q.  Do you think that would be a suitable resolution?  Anything more complicated will likely test my limited C coding skills...  5:S

Reply:
@ordex ENABLE_VLAN_TAGGING is completely optional, which is why this patch activates it by default (AFAIR). "do we still have the 802.1q header in the packet?" No, nothing changes without specifying `--vlan-tagging` in the server configuration.

Reply:
That's correct, I don't think any packets are ever modified, but it looks like the vid read there does get processed later on by [multi_bcast](https://github.com/OpenVPN/openvpn/pull/76/files/d9ee9f365ea49b9b843f06a36cf4e5249ca9798e#diff-2ef74147f86cc5c8bac60c4ebc16a713R2260), and if the packet had no 802.1q header, the vid would be set to the second byte of the payload.  I'm not 100% certain of the codepath before or after it, so there may well be other mitigations in place, but if not then that might be a security issue, potentially allowing for vlan hopping.  Ideally, the vid would not be read if the packet doesn't contain an 802.1q header (and 0 used instead?).  I think if `vlan-tagging` were not set it might just lead to dropped packets.  I also realized that the solution I suggested earlier wouldn't help in situations where `ENABLE_PF` wasn't defined.

Reply:
Yeah, I meant what @ikelos : the VID extraction seems to be performed even when `--vlan-tagging` is not enabled, without checking if the header is actially 802.1q (unless I overlooked something - I still have to go through all the code).

thanks both for bearing with me

Reply:
(Ah sorry, only read ordex's comment and not the referenced code.)
`mroute_copy_ether_to_addr` is concerned with *extracting* the MAC/ethernet address from the packet and storing it in mroute_addr. In case vlan tagging support is compiled in, it also stores the VID (passed in by parameter) in mroute_addr.
Any suggestions for a better function name?

Reply:
I think the function name is ok, but storing the vid when the packet isn't an 802.1q packet is where the issue arises.  A conditional to check that the packet does actually contain a vid would resolve the problem...

Reply:
@ikelos The function does not read the vid from the packet, only the ethernet address. The ethernet address is then written to maddr. When ENABLE_VLAN_TAGGING is compiled in, it also writes the vid parameter into maddr. This is independent of whether the packet is 802.1q or not. (If we don't write a vid into maddr, that part of maddr would potentially stay uninitialized.)

So at least for this part of the code I don't see any potential issues. But maybe I just don't get what you're saying?

Reply:
@fknittel My bad, sorry about the confusion, as I said my C skills are pretty poor.  I'd thought it was memcpy(src, dest, count), not memcpy(dest, src, count).

Reply:
My unfinished-december-2016-rebase contains the following approach:
```
        struct {
            uint8_t mac[OPENVPN_ETH_ALEN];
#ifdef ENABLE_VLAN_TAGGING
            int16_t vid;
#endif
        } eth_addr;
```

Replacing `mroute_addr.eth_addr` with `mroute_addr.eth_addr.mac`. Results in something like this:

```
static void
mroute_copy_ether_to_addr(struct mroute_addr *maddr,
                          const uint8_t *eth_addr,
                          uint16_t vid)
{
    maddr->type = MR_ADDR_ETHER;
    maddr->netbits = 0;
    memcpy(maddr->eth_addr.mac, eth_addr, sizeof(maddr->eth_addr.mac));
#ifdef ENABLE_VLAN_TAGGING
    maddr->eth_addr.vid = vid;
#endif
    maddr->len = sizeof(maddr->eth_addr);
}
```

Reply:
@fknittel Cool, do you want me to integrate that into my branch or would you prefer to do it (I'm fine either way)?  Are there other things from your unfinished-december-2016-rebase I should pull across?  I'm happy for you to commit/rebase in my branch if that'll help us all work in one place?